R/quick-theme.R

```diff
@@ -0,0 +1,210 @@
+#' `$gray-900` variables; or on the theme colors (primary, secondary, danger,
+#' warning, info, etc.). `bs_theme_quick` sets `$white` to the `bg` color,
+#' `$black` to the `fg` color, and interpolates the grays between them. If
+#' provided, the `accent` argument is used to set the `$primary` variable, and
+#' the `secondary` argument is used to set `$secondary` and `$default`
+#' variables.
+#'
+#' For Bootstrap 3, a similar set of `$black`, `$white`, and `$gray-darker`
+#' through `$gray-lighter` variables exist, and these are populated using the
+#' same strategy as with Bootstrap 4. However, unlike Bootstrap 4, in Bootstrap
+#' 3 many of the default colors are hard-coded hex colors, that also happen to
+#' be shades of gray. `bs_theme_quick` overrides these hard-coded values with
+#' colors interpolated between `bg` and `fg`.
+#'
+#' @examples
+#'
+#' bs_theme_new("4+3")
+#' bs_theme_quick(bg = "#000060", fg = "skyblue",
+#'   accent = "orange", secondary = "silver")
+#'
+#' # You can apply further customizations here if desired, e.g.:
+#' bs_theme_add_variables("success" = "#1D7732")
+#'
+#' if (interactive()) {
+#'   bs_theme_preview()
+#' }
+#'
+#' @export
+bs_theme_quick <- function(bg = "#FFFFFF", fg = "#000000",
+  accent = NULL, secondary = NULL) {
+
+  theme <- bs_theme_get()
+  if (is.null(theme)) {
+    stop("No bootstraplib theme is active (did you forget to call bs_theme_new()?)")
+  }
+
+  bs4_tags <- c("bootstraplib_version_4", "bootstraplib_version_4+3")
+  bs3_tags <- c("bootstraplib_version_3")
+
+  results <- if (any(bs4_tags %in% theme$tags)) {
+    bs4_theme_quick(bg, fg, accent, secondary)
+  } else if (any(bs3_tags %in% theme$tags)) {
+    bs3_theme_quick(bg, fg, accent, secondary)
+  } else {
+    stop("bs_theme_quick doesn't recognize the active version of Bootstrap")
+  }
+  bs_theme_add(results)
+}
+
+bs4_theme_quick <- function(bg, fg, accent, secondary) {
+  white <- htmltools:::parseCssColors(bg)
+  black <- htmltools:::parseCssColors(fg)
+
+  grays <- colorRamp(c(white, black), alpha = TRUE)(0:10/10)
+
+  if (any(grays[,4] != 255)) {
+    warning(call. = FALSE,
+      "bs_theme_quick does not respect alpha in `white` and `black` arguments"
+    )
+  }
+
+  white_yiq <- color_yiq(grays[1,1], grays[1,2], grays[1,3])
+  black_yiq <- color_yiq(grays[11,1], grays[11,2], grays[11,3])
+
+  grays <- sprintf("#%02X%02X%02X",
+    round(grays[,1]),
+    round(grays[,2]),
+    round(grays[,3])
+  )
+
+  names(grays) <- c(
+    "white",
+    paste0("gray-", 1:9 * 100),
+    "black"
+  )
+
+  results <- as.list(grays)
+
+  if (!is.null(accent)) {
+    results <- c(results, list(primary = accent))
+  }
+  if (!is.null(secondary)) {
+    results <- c(results, list(
+      secondary = secondary,
+      default = secondary
+    ))
+  }
+  if (white_yiq < black_yiq) {
+    # Invert yiq colors
+    results <- c(results, list(
+      "yiq-text-light" = grays[["gray-900"]],
+      "yiq-text-dark" = grays[["white"]]
+    ))
+  }
+
```

Comment:
Add `!default` flags here

Reply:
I don't think it's needed? This list is going into `sass_layer(defaults = ...)`

Reply:
`sass_layer(defaults = ...)` won't automatically add a `!default` flag...only `bs_theme_add_variables()` will at the moment

Reply:
Oooh good catch then. I fixed and added a test.